PROD-2049: Implement sending emails via property-specific messaging templates

eastandwestwind opened the pull request:
Closes PROD-2049

### Description Of Changes

This change adds the core logic to determine both:

1. Whether or not to use property-specific messaging templates
2. Which messaging template to use when sending emails from Fides

We support property-specific messaging for the following message types
- [ ] PRIVACY_REQUEST_REVIEW_APPROVE
- [ ] PRIVACY_REQUEST_REVIEW_DENY
- [ ] PRIVACY_REQUEST_RECEIPT
- [ ] PRIVACY_REQUEST_COMPLETE_ACCESS
- [ ] PRIVACY_REQUEST_COMPLETE_DELETION
- [ ] SUBJECT_IDENTITY_VERIFICATION (both consent and privacy request)


### Code Changes

* [ ] Adds new env var to track whether or not we enable property-specific messaging
* [ ] Adds `property_id` field to `PrivacyRequest` model 
* [ ] Implements retrieving appropriate messaging template by property id if ENV var is enabled

### Steps to Confirm

* [ ] With `FIDES__NOTIFICATIONS__ENABLE_PROPERTY_SPECIFIC_MESSAGING` disabled, confirm that no regressions exist with configuring email templates and sending out emails via Fides (consult the above list for email types to test)
* [ ] With `FIDES__NOTIFICATIONS__ENABLE_PROPERTY_SPECIFIC_MESSAGING` enabled, confirm that we can send property-specific emails. To do this, to the following:
* [ ] Configure your email messaging config (e.g. Mailgun) as you did before to enable sending emails from Fides
* [ ] Enable your property-specific emails. In the postgres container, run `update messaging_template set is_enabled=true;`
 

### Pre-Merge Checklist

* [ ] All CI Pipelines Succeeded
* Documentation:
  * [ ] documentation complete, [PR opened in fidesdocs](https://github.com/ethyca/fidesdocs/pulls)
  * [ ] documentation [issue created in fidesdocs](https://github.com/ethyca/fidesdocs/issues/new/choose)
  * [ ] if there are any new client scopes created as part of the pull request, remember to update public-facing documentation that references our scope registry
* [ ] Issue Requirements are Met
* [ ] Relevant Follow-Up Issues Created
* [x] Update `CHANGELOG.md`
* [ ] For API changes, the [Postman collection](https://github.com/ethyca/fides/blob/main/docs/fides/docs/development/postman/Fides.postman_collection.json) has been updated
* If there are any database migrations:
  * [ ] Ensure that your downrev is up to date with the latest revision on `main`
  * [ ] Ensure that your `downgrade()` migration is correct and works
    * [ ] If a downgrade migration is not possible for this change, please call this out in the PR description!

eastandwestwind commented on src/fides/api/service/messaging/message_dispatch_service.py:
```diff
@@ -151,11 +216,24 @@ def dispatch_message(
+    # If property-specific messaging is enabled, we switch over to this mode, regardless of other ENV vars
+    if ConfigProxy(db).notifications.enable_property_specific_messaging:
+        if not property_specific_messaging_template:
+            logger.warning(
+                "Skipping sending property-specific email as no enabled template was found for action type: {}",
+                action_type,
+            )
```
The current product need is that a user can enable / disable the templates themselves in the admin-ui. So this means if we do not find any "enabled" templates linked to the property and template type, we do not send the email. Basically:

- "Disabled" == do not send email for that template type, for the properties linked to that template.

That said, we can still attempt to handle sending default emails for properties that are NOT explicitly linked to a template. See #7 here- https://ethyca.atlassian.net/wiki/spaces/EN/pages/3055681577/2024-05-09+Property+Specific+Messaging#Open-questions%3A. We need some product feedback here.

galvana commented on src/fides/api/service/messaging/messaging_crud_service.py:
```diff
@@ -110,10 +110,39 @@ def get_all_basic_messaging_templates(db: Session) -> List[MessagingTemplate]:
+                    (MessagingTemplate.type == template_type)
+                    & (Property.id == default_property.id)
+                ),
+            ).first()
```
Sorry! You can ignore this, my comment doesn't apply to what you have now

galvana commented:
Everything worked with `FIDES__NOTIFICATIONS__ENABLE_PROPERTY_SPECIFIC_MESSAGING` disabled but I ran into this serialization error when I set it to true and tried to use the default templates.

```
fides  |   File "/fides/src/fides/api/api/v1/endpoints/privacy_request_endpoints.py", line 1131, in verify_identification_code
fides  |     _send_privacy_request_receipt_message_to_user(
fides  |   File "/fides/src/fides/api/api/v1/endpoints/privacy_request_endpoints.py", line 257, in _send_privacy_request_receipt_message_to_user
fides  |     dispatch_message_task.apply_async(
fides  |   File "/opt/fides/lib/python3.10/site-packages/celery/app/task.py", line 564, in apply_async
fides  |     content_type, content_encoding, data = serialization.dumps(
fides  |   File "/opt/fides/lib/python3.10/site-packages/kombu/serialization.py", line 219, in dumps
fides  |     with _reraise_errors(EncodeError):
fides  |   File "/usr/local/lib/python3.10/contextlib.py", line 153, in __exit__
fides  |     self.gen.throw(typ, value, traceback)
fides  |   File "/opt/fides/lib/python3.10/site-packages/kombu/serialization.py", line 45, in _reraise_errors
fides  |     reraise(wrapper, wrapper(exc), sys.exc_info()[2])
fides  |   File "/opt/fides/lib/python3.10/site-packages/kombu/exceptions.py", line 34, in reraise
fides  |     raise value.with_traceback(tb)
fides  |   File "/opt/fides/lib/python3.10/site-packages/kombu/serialization.py", line 41, in _reraise_errors
fides  |     yield
fides  |   File "/opt/fides/lib/python3.10/site-packages/kombu/serialization.py", line 220, in dumps
fides  |     payload = encoder(data)
fides  |   File "/opt/fides/lib/python3.10/site-packages/kombu/utils/json.py", line 63, in dumps
fides  |     return _dumps(s, cls=cls, **dict(default_kwargs, **kwargs))
fides  |   File "/usr/local/lib/python3.10/json/__init__.py", line 238, in dumps
fides  |     **kw).encode(obj)
fides  |   File "/usr/local/lib/python3.10/json/encoder.py", line 199, in encode
fides  |     chunks = self.iterencode(o, _one_shot=True)
fides  |   File "/usr/local/lib/python3.10/json/encoder.py", line 257, in iterencode
fides  |     return _iterencode(o, 0)
fides  |   File "/opt/fides/lib/python3.10/site-packages/kombu/utils/json.py", line 47, in default
fides  |     return super().default(o)
fides  |   File "/usr/local/lib/python3.10/json/encoder.py", line 179, in default
fides  |     raise TypeError(f'Object of type {o.__class__.__name__} '
fides  | kombu.exceptions.EncodeError: Object of type MessagingTemplate is not JSON serializable
```

galvana reviewed (commented):
Overall this looks good! Just a few things.

1. I left a comment on the expected behavior between properties and privacy requests and I'd like to hear your thoughts.
2. I got an error when running with `FIDES__NOTIFICATIONS__ENABLE_PROPERTY_SPECIFIC_MESSAGING=true`, I left a root-level comment with the stacktrace
3. I had to enable the default templates in order to get messaging to work. Is it expected that they're set to disabled by default? I imagine we would want them enabled for OSS users?

eastandwestwind commented on src/fides/api/alembic/migrations/versions/a3c173391603_add_property_id_to_privacy_request_model.py:
```diff
@@ -0,0 +1,41 @@
+    op.add_column(
+        "privacyrequest", sa.Column("property_id", sa.String(), nullable=True)
+    )
+    op.create_foreign_key(
+        "privacyrequest_plus_property",
+        "privacyrequest",
+        "plus_property",
+        ["property_id"],
+        ["id"],
+    )
```
@galvana good point, seems like we should instead not have the foreign key so properties can be delete freely here 👍

eastandwestwind commented on src/fides/api/api/v1/endpoints/privacy_request_endpoints.py:
```diff
@@ -1774,6 +1822,7 @@ def create_privacy_request_func(
 
     If authenticated is True the identity verification step is bypassed.
     """
+    # TODO: (PROD-2142)- update privacy center to pass in property id where applicable
```
I've added verification in 2 places- one in DSR create, and one in consent request create 👍

eastandwestwind commented:
Thanks so much for the review @galvana !

1. Updated as per your suggestion, including adding validation for valid property_id, if passed in to both consent and DSR requests
2. Fixed the error- turns out I can't pass in DB models to celery workers.
3. Yes, you'll need to enable the default templates first, I've updated the testing instructions to reflect this. We don't support property-specific messaging in OSS at all. This is only a Plus feature.

galvana commented on src/fides/api/service/messaging/message_dispatch_service.py:
```diff
@@ -221,6 +244,9 @@ def dispatch_message(
 
     # If property-specific messaging is enabled, we switch over to this mode, regardless of other ENV vars
     if ConfigProxy(db).notifications.enable_property_specific_messaging:
+        property_specific_messaging_template = get_property_specific_messaging_template(
+            db, property_id, action_type
+        )
```
I like how this refactor moved this call to one location 💯

galvana reviewed (approved):
Thanks for making the changes @eastandwestwind!

> Fixed the error- turns out I can't pass in DB models to celery workers.

Thanks for the fix here! I re-ran this scenario and it worked as expected

> Yes, you'll need to enable the default templates first, I've updated the testing instructions to reflect this. We don't support property-specific messaging in OSS at all. This is only a Plus feature.

I was thinking of it wrong, this makes sense!
